data/transactions/application.go

```diff
@@ -665,6 +665,12 @@ func (ac *ApplicationCallTxnFields) apply(header Header, balances Balances, spec
 		return err
 	}
 
+	// Ensure that the only operation we can do is ClearState if the application
+	// does not exist
+	if !exists && ac.OnCompletion != ClearStateOC {
+		return fmt.Errorf("only clearing out is supported for applications that do not exist")
+	}
+
```

Comment:
I wonder if you can just completely handle a deleted application at this point and instantly return (this might involve moving some chunks of code around).